macOS build regression: wal.cpp uses Linux-only fdatasync()

## Summary

Commit \`1723f819e\` ("WAL durability: POSIX fd + fdatasync, framed
entries with checksum") introduced unconditional calls to \`::fdatasync(int)\`
in \`src/storage/wal.cpp\`. On macOS (\`build-portable\`, AppleClang 15.0)
the build fails because \`fdatasync\` is Linux-specific and is not
declared by macOS's \`<unistd.h>\`.

## Reproduction

\`\`\`bash
git checkout 99a1a1250   # current turbolynx-dslab/main tip
cmake -GNinja -DCMAKE_BUILD_TYPE=Release -DBUILD_PYTHON=ON \
      -DTURBOLYNX_PORTABLE_DISK_IO=ON -DENABLE_TCMALLOC=OFF \
      -DBUILD_UNITTESTS=OFF -DTBB_TEST=OFF \
      -B build-portable
cmake --build build-portable
\`\`\`

## Error

\`\`\`
src/storage/wal.cpp:107:11: error: no member named 'fdatasync' in the global namespace
src/storage/wal.cpp:213:11: error: no member named 'fdatasync' in the global namespace
src/storage/wal.cpp:347:11: error: no member named 'fdatasync' in the global namespace
\`\`\`

Three call sites, each wrapped by a macOS-incompatible branch:

- Line 107 — fdatasync after writing the WAL header
- Line 213 — fdatasync after each framed log entry
- Line 347 — fdatasync in the Flush path

## Environment

- macOS darwin 24.6.0 (arm64, M-series)
- AppleClang 15.0
- CMake 4.3.0, Ninja 1.13.0
- Commit \`99a1a1250\` (\`turbolynx-dslab/main\` tip)
- CMake flags: default macOS portable build (\`-DTURBOLYNX_PORTABLE_DISK_IO=ON\`)

## Suggested direction

Two idiomatic options for the macOS fallback:

1. **\`fsync(fd)\` on macOS** — simpler, flushes both data and metadata
   (slightly more work than \`fdatasync\`, but semantically safe).
2. **\`fcntl(fd, F_FULLFSYNC)\`** — macOS's strict "flush to platter"
   equivalent; more expensive but matches the durability intent of the
   \`fdatasync\` call sites.

A thin inline helper, e.g.

\`\`\`cpp
// src/include/storage/wal.hpp or similar
inline int wal_datasync(int fd) {
#if defined(__APPLE__)
    return ::fsync(fd);             // or fcntl(fd, F_FULLFSYNC)
#else
    return ::fdatasync(fd);
#endif
}
\`\`\`

and replacing the three \`::fdatasync(fd_)\` call sites with \`wal_datasync(fd_)\`
would restore the macOS build without changing Linux semantics. Happy to
open a PR if the team would prefer that.

## Impact

Blocks \`build-portable\` on macOS at HEAD. The \`oss-supply-chain\`
application's pytest suite cannot run against the latest engine until
this is resolved.

--- a/src/include/common/fdatasync.hpp
+++ b/src/include/common/fdatasync.hpp
@@ -1,0 +1,28 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// common/fdatasync.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <unistd.h>
+
+namespace duckdb {
+
+// Cross-platform equivalent of POSIX fdatasync(2). On macOS, where
+// fdatasync is not declared, falls back to fsync(2). Note that on macOS
+// this only flushes the OS page cache to the drive — it does NOT flush
+// the drive's own write cache. Strict "data on media" durability on
+// macOS additionally requires fcntl(fd, F_FULLFSYNC); tracked separately.
+inline int Fdatasync(int fd) {
+#if defined(__APPLE__)
+    return ::fsync(fd);
+#else
+    return ::fdatasync(fd);
+#endif
+}
+
+} // namespace duckdb

--- a/src/storage/wal.cpp
+++ b/src/storage/wal.cpp
@@ -10,6 +10,7 @@
 #include "storage/delta_store.hpp"
 #include "common/checksum.hpp"
 #include "common/exception.hpp"
+#include "common/fdatasync.hpp"
 #include "spdlog/spdlog.h"
 
 #include <cerrno>
@@ -104,7 +105,7 @@ void WALWriter::WriteHeader() {
         p += n;
         remaining -= (size_t)n;
     }
-    if (::fdatasync(fd_) != 0) {
+    if (Fdatasync(fd_) != 0) {
         int err = errno;
         if (err == EIO) {
             throw FatalException("[WAL] fdatasync(%s) returned EIO after header", path_);
@@ -210,7 +211,7 @@ void WALWriter::CommitEntry() {
     //    metadata we don't need (mtime), so it's cheaper than fsync. EIO
     //    from fsync is unrecoverable — per the fsyncgate discussion the
     //    dirty pages are gone and retrying won't help.
-    if (::fdatasync(fd_) != 0) {
+    if (Fdatasync(fd_) != 0) {
         int err = errno;
         if (err == EIO) {
             throw FatalException("[WAL] fdatasync(%s) returned EIO — data may be lost", path_);
@@ -344,7 +345,7 @@ void WALWriter::Flush() {
     // common case. Keep it as an explicit sync for defensive callers and in
     // case future code adds unsynced writes.
     if (fd_ < 0) return;
-    if (::fdatasync(fd_) != 0) {
+    if (Fdatasync(fd_) != 0) {
         int err = errno;
         if (err == EIO) {
             throw FatalException("[WAL] fdatasync(%s) returned EIO", path_);

